core/src/main/java/io/questdb/cairo/ConcurrentBitmapIndexFwdReader.java

```diff
@@ -0,0 +1,226 @@
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo;
+
+import io.questdb.cairo.sql.RowCursor;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.Unsafe;
+import io.questdb.std.str.Path;
+
+/**
+ * Cursors returned by this class are thread-safe since they never extend key and
+ * value memories or mutate any shared variables.
+ * <p>
+ * The important assumption is that of() method should be called on a single thread
+ * prior to any cursor operations. Once this happens, cursors can be obtained and
+ * called concurrently.
+ */
+public class ConcurrentBitmapIndexFwdReader extends AbstractIndexReader {
+    private final static Log LOG = LogFactory.getLog(ConcurrentBitmapIndexFwdReader.class);
+    private final Cursor cursor = new Cursor();
+
+    public ConcurrentBitmapIndexFwdReader() {
+    }
+
+    // test only
+    public ConcurrentBitmapIndexFwdReader(
+            CairoConfiguration configuration,
+            Path path,
+            CharSequence name,
+            long columnNameTxn,
+            long unIndexedNullCount
+    ) {
+        of(configuration, path, name, columnNameTxn, unIndexedNullCount, -1);
+    }
+
+    @Override
+    public RowCursor getCursor(boolean cachedInstance, int key, long minValue, long maxValue) {
+        if (key == 0 && unIndexedNullCount > 0 && minValue < unIndexedNullCount) {
+            // we need to return some nulls and the whole set of actual index values
+            final Cursor cursor = getCursor(cachedInstance);
+            cursor.of(key, 0, maxValue, keyCount, minValue, unIndexedNullCount);
+            return cursor;
+        }
+
+        if (key < keyCount) {
+            final Cursor cursor = getCursor(cachedInstance);
+            cursor.of(key, minValue, maxValue, keyCount, 0, 0);
+            return cursor;
+        }
+
+        return EmptyRowCursor.INSTANCE;
+    }
+
+    /**
+     * Allows reusing cursor objects, if that's possible.
+     */
+    public RowCursor initCursor(RowCursor rowCursor, int key, long minValue, long maxValue) {
+        Cursor cursor = null;
+        if (rowCursor != null && rowCursor != EmptyRowCursor.INSTANCE) {
+            cursor = (Cursor) rowCursor;
+            assert cursor.parentReader() == this;
+        }
+
+        if (key == 0 && unIndexedNullCount > 0 && minValue < unIndexedNullCount) {
+            // we need to return some nulls and the whole set of actual index values
+            if (cursor == null) {
+                cursor = getCursor(false);
+            }
+            cursor.of(key, 0, maxValue, keyCount, minValue, unIndexedNullCount);
+            return cursor;
+        }
+
+        if (key < keyCount) {
+            if (cursor == null) {
+                cursor = getCursor(false);
+            }
+            cursor.of(key, minValue, maxValue, keyCount, 0, 0);
+            return cursor;
+        }
+
+        return EmptyRowCursor.INSTANCE;
```

Comment:
so is this

Reply:
Done in 9cc86ee6e93baaf2ac4b7207b45d486d66dccf07